weather_mv/loader_pipeline/ee.py

```diff
@@ -0,0 +1,371 @@
+
+        if not re.match("^projects/.+/assets.*", known_args.ee_asset):
+            raise RuntimeError("'--ee_asset' is required to be in format: projects/+/assets/*.")
+
+        # Check that Cloud resource regions are consistent.
+        if not (known_args.dry_run or known_args.skip_region_validation):
+            # Program execution will terminate on failure of region validation.
+            logger.info('Validating regions for data migration. This might take a few seconds...')
+            validate_region(temp_location=pipeline_options_dict.get('temp_location'),
+                            region=pipeline_options_dict.get('region'))
+            logger.info('Region validation completed successfully.')
+
+    def expand(self, paths):
+        """Converts gribs from paths into tiff and uploads them into the earth engine."""
+        if not self.dry_run:
+            (
+                paths
+                | 'ConvertToCog' >> beam.ParDo(
+                    ConvertToCog(
+                        tiff_location=self.tiff_location,
+                        open_dataset_kwargs=self.xarray_open_dataset_kwargs,
+                        disable_in_memory_copy=self.disable_in_memory_copy,
+                        disable_grib_schema_normalization=self.disable_grib_schema_normalization))
+                | 'IngestIntoEE' >> beam.ParDo(
+                    IngestIntoEE(
+                        ee_asset=self.ee_asset,
+                        use_personal_account=self.use_personal_account))
+            )
+        else:
+            (
+                paths
+                | 'Log Grib Files' >> beam.Map(logger.debug)
+            )
+
+
+@dataclasses.dataclass
+class ConvertToCog(beam.DoFn):
+    """Writes tiff image after extracting grib data and uploads it to GCS.
+
+    Attributes:
+        tiff_location: The bucket location at which tiff files will be pushed.
+        open_dataset_kwargs: A dictionary of kwargs to pass to xr.open_dataset().
+        disable_in_memory_copy: A flag to turn in-memory copy off; Default: on.
+        disable_grib_schema_normalization: A flag to turn grib schema normalization off; Default: on.
+    """
+
+    tiff_location: str
+    open_dataset_kwargs: t.Optional[t.Dict] = None
+    disable_in_memory_copy: bool = False
+    disable_grib_schema_normalization: bool = False
+
+    def _get_tiff_name(self, uri: str) -> str:
+        """Extracts file name and converts it into an EE-safe name"""
+        file_name = os.path.basename(uri)
+        base_name = os.path.splitext(file_name)[0]
+        tiff_name = base_name.replace('.', '_')
+        return tiff_name
+
+    def _get_target_path(self, file_name: str) -> str:
+        """Creates a GCS target path from the file name."""
+        return os.path.join(self.tiff_location, file_name)
+
+    def process(self, uri: str) -> t.Iterator[TiffData]:
+        """Opens grib files and yields TiffData."""
+
+        logger.info(f'Converting {uri!r} to COGs ...')
```

Comment:
Oh I added the dry run check in ToEarthEngine's expand. It will just log these path uris and won't go inside any transforms. WDYT? Should we let it go inside each transform in case of dry run?

```
def expand(self, paths):
        """Converts gribs from paths into tiff and uploads them into the earth engine."""
        if not self.dry_run:
            (
                paths
                | 'ConvertToCog' >> beam.ParDo(
                    ConvertToCog(
                        tiff_location=self.tiff_location,
                        open_dataset_kwargs=self.xarray_open_dataset_kwargs,
                        disable_in_memory_copy=self.disable_in_memory_copy,
                        disable_grib_schema_normalization=self.disable_grib_schema_normalization))
                | 'IngestIntoEE' >> beam.ParDo(
                    IngestIntoEE(
                        ee_asset=self.ee_asset,
                        use_personal_account=self.use_personal_account))
            )
        else:
            (
                paths
                | 'Log Grib Files' >> beam.Map(logger.debug)
            )
```